### PR TITLE
ICU-21410 fix int32_t overflow in listFormat

### PR DESCRIPTION
**Pull request by FrankYFTang**
##### Checklist

- [X] Issue filed: https://unicode-org.atlassian.net/browse/ICU-21410
- [X] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [X] Tests included
- [X] Documentation is changed or added

### CODE REVIEW

File: icu4c/source/i18n/formatted_string_builder.cpp [hunk 1]
```diff
@@ -152,6 +154,7 @@ FormattedStringBuilder::insertCodePoint(int32_t index, UChar32 codePoint, Field
     if (U_FAILURE(status)) {
         return count;
     }
+    U_ASSERT(position >= 0);
```
Comment (sffc) [inline icu4c/source/i18n/formatted_string_builder.cpp:157]: Can you put all these `U_ASSERT`s inside prepareForInsert?

File: icu4c/source/i18n/formatted_string_builder.cpp [hunk 2]
```diff
@@ -276,6 +283,12 @@ int32_t FormattedStringBuilder::prepareForInsertHelper(int32_t index, int32_t co
     char16_t *oldChars = getCharPtr();
     Field *oldFields = getFieldPtr();
     if (fLength + count > oldCapacity) {
+        if ((fLength + count) > std::numeric_limits<int32_t>::max() / 2) {
```
Comment (sffc) [inline icu4c/source/i18n/formatted_string_builder.cpp:286]: You can use `INT32_MAX` instead in ICU4C

**Review by sffc — CHANGES_REQUESTED** [review]

**Comment by FrankYFTang** [comment]
PTAL

**Review by sffc — APPROVED** [review]

**Comment by sffc** [comment]
LGTM please squash

**Comment by FrankYFTang** [comment]
squashed , please approve again

**Review by sffc — APPROVED** [review]

**Pull request merged**